solvebio/cli/data.py

```diff
@@ -159,10 +162,63 @@ def _upload_folder(
                     )
                 )
             else:
-                remote_parent = Object.get_by_full_path(
-                    remote_folder_full_path, assert_type="folder"
-                )
-                Object.upload_file(local_file_path, remote_parent.path, vault.full_path)
+                all_files.append((local_file_path, remote_folder_full_path, vault.full_path))
+
+    # Identify which remote folders already exist
+    upload_root_path, _ = Object.validate_full_path(
+        os.path.join(base_remote_path, local_start)
+    )
+    results = GlobalSearch().filter(path__prefix=upload_root_path, type="folder")
+    remote_folders_existing = set([x.full_path for x in results])
+    all_folder_parts = set()
+    for vault, remote_folder_path in all_folders:
+        subfolders = remote_folder_path.lstrip("/").split("/")
+        parent_folder_path = ""
+        for folder in subfolders:
+            folder_full_path = os.path.join(parent_folder_path, folder)
+            if folder_full_path not in remote_folders_existing and not folder_full_path.endswith(":"):
+                all_folder_parts.add(folder_full_path)
+            parent_folder_path = folder_full_path
+
+    if not dry_run:
+        # Create folders serially since these require
+        # previous folders to be created so that parent_object_id can
+        # be populated
+        all_folder_parts = sorted(all_folder_parts, key=lambda x: len(x.split("/")))
+        for folder in all_folder_parts:
+            _create_folder(vault, folder)
+
+        # Create files in parallel
+        pool = multiprocessing.Pool(num_processes)
+        for result in pool.imap_unordered(_create_file_job, all_files):
+            # Check if an exception was raised by the pool worker
+            # which is returned at the end of the tuple
+            if len(result) == 4:
```

Comment:
Does upload_file retry at all in the event of (internal) failure?

Reply:
Yup, I think under `solvebio.resource.object.Object.upload_file` there's a max_retries set to 5 on the s3 upload URL.
Under `solvebio.client.SolveClient.__init__` all EDP api requests also have a total of 5 retries.


What do you mean by internal failure? Usually retries are triggered due to external failures.

Reply:
Ok great, I just meant internal within the upload file function itself.

Reply:
Still not sure what you mean by this. Is there anything we want to see changed here?

Reply:
No worries I think it's all good. Was just wondering if the underlying upload function had retry logic built-in so that we don't need additional logic outside in the even that one of the multiprocessing tasks fails.